### PR TITLE
Fixes #276 - attach appengine specific jar to app engine facet

### LINKED ISSUE #276: attach appengine specific jar to app engine facet

### PR DESCRIPTION
**Pull request by akerekes**

### CODE REVIEW

File: plugins/com.google.cloud.tools.eclipse.appengine.facets/src/com/google/cloud/tools/eclipse/appengine/sdk/AppEngineSDKClasspathContainer.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
+package com.google.cloud.tools.eclipse.appengine.sdk;
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.facets/src/com/google/cloud/tools/eclipse/appengine/sdk/AppEngineSDKClasspathContainer.java]: I'm not sure a new package is called for here. Any reason not to just put it in the facet pckage?

File: plugins/com.google.cloud.tools.eclipse.appengine.facets/src/com/google/cloud/tools/eclipse/appengine/sdk/AppEngineSDKClasspathContainer.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+import org.eclipse.core.runtime.IPath;
+import org.eclipse.core.runtime.Path;
+import org.eclipse.jdt.core.IClasspathContainer;
+import org.eclipse.jdt.core.IClasspathEntry;
+import org.eclipse.jdt.core.JavaCore;
+
+import com.google.cloud.tools.appengine.cloudsdk.PathResolver;
+
+public final class AppEngineSDKClasspathContainer implements IClasspathContainer {
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.facets/src/com/google/cloud/tools/eclipse/appengine/sdk/AppEngineSDKClasspathContainer.java]: Gogole style is Sdk (no internal caps)

File: plugins/com.google.cloud.tools.eclipse.appengine.facets/src/com/google/cloud/tools/eclipse/appengine/sdk/AppEngineSDKClasspathContainer.java [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+  private static final String SDK_JAR = "/platform/google_appengine/google/appengine/tools/java/lib/appengine-tools-api.jar";
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.facets/src/com/google/cloud/tools/eclipse/appengine/sdk/AppEngineSDKClasspathContainer.java]: soon, if not yet, we can all exceot the last piece form app-tools-lib. Please add a todo.

File: plugins/com.google.cloud.tools.eclipse.appengine.facets/src/com/google/cloud/tools/eclipse/appengine/sdk/AppEngineSDKClasspathContainer.java [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+  @Override
+  public String getDescription() {
+    return "App Engine SDKs";
+  }
+
+  @Override
+  public IClasspathEntry[] getClasspathEntries() {
+    java.nio.file.Path cloudSdkPath = PathResolver.INSTANCE.getCloudSdkPath();
+    if (cloudSdkPath != null) {
+      String appEngineToolsApiJar = cloudSdkPath + SDK_JAR;
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.facets/src/com/google/cloud/tools/eclipse/appengine/sdk/AppEngineSDKClasspathContainer.java]: should use resolve() method here from java.nio.file.Path to better handle platform differences on windows.

File: plugins/com.google.cloud.tools.eclipse.appengine.facets/src/com/google/cloud/tools/eclipse/appengine/sdk/AppEngineSDKClasspathContainer.java (hunk 2 above)
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.facets/src/com/google/cloud/tools/eclipse/appengine/sdk/AppEngineSDKClasspathContainer.java]: Th8is class could have some tests.

**Comment by akerekes** [comment]
@briandealwis @nbashirbello PTAL

File: plugins/com.google.cloud.tools.eclipse.appengine.facets.test/src/com/google/cloud/tools/eclipse/appengine/facets/AppEngineSdkClasspathContainerTest.java [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+import org.eclipse.core.runtime.IPath;
+import org.eclipse.jdt.core.IClasspathEntry;
+import org.junit.Test;
+
+public class AppEngineSdkClasspathContainerTest {
+
+  @Test
+  public void testGetPath() {
+    IPath path = new AppEngineSdkClasspathContainer().getPath();
+    assertThat(path.segmentCount(), is(1));
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.facets.test/src/com/google/cloud/tools/eclipse/appengine/facets/AppEngineSdkClasspathContainerTest.java]: I thought you could say something like 
assertThat(path.segmentCount()).is(1)?
If not, I think assertEquals is clearer

File: plugins/com.google.cloud.tools.eclipse.appengine.facets/src/com/google/cloud/tools/eclipse/appengine/facets/AppEngineSdkClasspathContainer.java [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+import org.eclipse.jdt.core.IClasspathEntry;
+import org.eclipse.jdt.core.JavaCore;
+
+import com.google.cloud.tools.appengine.cloudsdk.PathResolver;
+
+public final class AppEngineSdkClasspathContainer implements IClasspathContainer {
+
+  // TODO should be changed once app-tools-lib can provide the directory
+  // https://github.com/GoogleCloudPlatform/app-tools-lib-for-java/issues/134
+  private static final String SDK_JAR = "platform/google_appengine/google/appengine/tools/java/lib/appengine-tools-api.jar";
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.facets/src/com/google/cloud/tools/eclipse/appengine/facets/AppEngineSdkClasspathContainer.java]: should be available now in 0.1.2-SNAPSHOT

File: plugins/com.google.cloud.tools.eclipse.appengine.facets/src/com/google/cloud/tools/eclipse/appengine/facets/AppEngineSdkClasspathContainer.java [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+            JavaCore.newLibraryEntry(new Path(appEngineToolsApiJar),
+                                     null /* sourceAttachmentPath */,
+                                     null /* sourceAttachmentRootPath */);
+        return new IClasspathEntry[]{ appEngineToolsEntry };
+      }
+    }
+    return new IClasspathEntry[0];
+  }
+
+  // Visible for testing
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.facets/src/com/google/cloud/tools/eclipse/appengine/facets/AppEngineSdkClasspathContainer.java]: I think Guava, which we already depends on, exposes @VisibleForTesting

File: plugins/com.google.cloud.tools.eclipse.appengine.facets.test/src/com/google/cloud/tools/eclipse/appengine/facets/AppEngineSdkClasspathContainerTest.java (hunk 5 above)
Comment (akerekes) [inline plugins/com.google.cloud.tools.eclipse.appengine.facets.test/src/com/google/cloud/tools/eclipse/appengine/facets/AppEngineSdkClasspathContainerTest.java]: It's not possible to use assertThat like that, but I still prefer it to assertEquals, as in the latter the arguments are swapped (assertEquals(expectedVal, actualVal)) which is less seamless to read than this.

File: plugins/com.google.cloud.tools.eclipse.appengine.facets/src/com/google/cloud/tools/eclipse/appengine/facets/AppEngineSdkClasspathContainer.java (hunk 6 above)
Comment (akerekes) [inline plugins/com.google.cloud.tools.eclipse.appengine.facets/src/com/google/cloud/tools/eclipse/appengine/facets/AppEngineSdkClasspathContainer.java]: Issue filed: https://github.com/GoogleCloudPlatform/gcloud-eclipse-tools/issues/305

File: plugins/com.google.cloud.tools.eclipse.appengine.facets/src/com/google/cloud/tools/eclipse/appengine/facets/AppEngineSdkClasspathContainer.java (hunk 7 above)
Comment (akerekes) [inline plugins/com.google.cloud.tools.eclipse.appengine.facets/src/com/google/cloud/tools/eclipse/appengine/facets/AppEngineSdkClasspathContainer.java]: done

File: plugins/com.google.cloud.tools.eclipse.appengine.facets/src/com/google/cloud/tools/eclipse/appengine/facets/AppEngineSdkClasspathContainer.java [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+import org.eclipse.jdt.core.IClasspathContainer;
+import org.eclipse.jdt.core.IClasspathEntry;
+import org.eclipse.jdt.core.JavaCore;
+
+import com.google.cloud.tools.appengine.cloudsdk.PathResolver;
+import com.google.common.annotations.VisibleForTesting;
+
+public final class AppEngineSdkClasspathContainer implements IClasspathContainer {
+
+  // TODO should be changed once app-tools-lib can provide the directory
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.facets/src/com/google/cloud/tools/eclipse/appengine/facets/AppEngineSdkClasspathContainer.java]: actually this is fixed. I forgot to close the bug. We just need to upgrade to 0.1.2.

**Comment by elharo** [comment]
LGTM

File: plugins/com.google.cloud.tools.eclipse.appengine.facets/src/com/google/cloud/tools/eclipse/appengine/facets/AppEngineSdkClasspathContainer.java [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+  @Override
+  public IClasspathEntry[] getClasspathEntries() {
+    java.nio.file.Path cloudSdkPath = pathResolver.getCloudSdkPath();
+    if (cloudSdkPath != null) {
+      File jarFile = cloudSdkPath.resolve(SDK_JAR).toFile();
+      if (jarFile.exists()) {
+        String appEngineToolsApiJar = jarFile.getPath();
+        IClasspathEntry appEngineToolsEntry =
+            JavaCore.newLibraryEntry(new Path(appEngineToolsApiJar),
+                                     null /* sourceAttachmentPath */,
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.facets/src/com/google/cloud/tools/eclipse/appengine/facets/AppEngineSdkClasspathContainer.java]: Another option is to create named local variables for these that are set to null.

**Pull request merged**